Sources/GRPCHTTP2Core/Client/Connection/LoadBalancers/Subchannel.swift

```diff
@@ -0,0 +1,591 @@
+extension Subchannel {
+  /// A stream of events which can happen to the subchannel.
+  var events: AsyncStream<Event> {
+    self.event.stream
+  }
+
+  /// Run the subchannel.
+  ///
+  /// Running the subchannel will attempt to maintain a connection to a remote endpoint. At times
+  /// the connection may be idle but it will reconnect on-demand when a stream is requested. If
+  /// connect attempts fail then the subchannel may progressively spend longer in a transient
+  /// failure state.
+  ///
+  /// Events and state changes can be observed via the ``events`` stream.
+  func run() async {
+    await withDiscardingTaskGroup { group in
+      for await input in self.input.stream {
+        switch input {
+        case .connect:
+          self.handleConnectInput(in: &group)
+        case .backedOff:
+          self.handleBackedOffInput(in: &group)
+        case .close:
+          self.handleCloseInput(in: &group)
+        case .handleConnectionEvent(let event):
+          self.handleConnectionEvent(event, in: &group)
+        }
+      }
+    }
+
+    // Once the task group is done, the event stream must be also be finished. In normal operation
+    // this is handled via other paths. For cancellation it must be finished explicitly.
+    if Task.isCancelled {
+      self.event.continuation.finish()
+    }
+  }
+
+  /// Initiate a connection attempt, if possible.
+  func connect() {
+    self.input.continuation.yield(.connect)
+  }
+
+  /// Initiates graceful shutdown, if possible.
+  func close() {
+    self.input.continuation.yield(.close)
+  }
+
+  /// Make a stream using the subchannel if it's ready.
+  ///
+  /// - Parameter descriptor: A descriptor of the method to create a stream for.
+  /// - Returns: The open stream.
+  func makeStream(
+    descriptor: MethodDescriptor,
+    options: CallOptions
+  ) async throws -> Connection.Stream {
+    let connection: Connection? = self.state.withLockedValue { state in
+      switch state {
+      case .notConnected, .connecting, .closing, .closed:
+        return nil
+      case .connected(let connected):
+        return connected.connection
+      }
+    }
+
+    guard let connection = connection else {
+      throw RPCError(code: .unavailable, message: "subchannel isn't ready")
+    }
+
+    return try await connection.makeStream(descriptor: descriptor, options: options)
+  }
+}
+
+@available(macOS 14.0, iOS 17.0, watchOS 10.0, tvOS 17.0, *)
+extension Subchannel {
+  private func handleConnectInput(in group: inout DiscardingTaskGroup) {
+    let connection = self.state.withLockedValue { state in
+      state.makeConnection(
+        to: self.endpoint.addresses,
+        using: self.connector,
+        backoff: self.backoff,
+        defaultCompression: self.defaultCompression,
+        enabledCompression: self.enabledCompression
+      )
+    }
+
+    guard let connection = connection else {
+      // Not in a state to start a connection.
```

Comment:
I don't think so: in this instance we're not even attempting the connect because we're either already trying or shutdown. If it's shutdown then the caller will know via the `shutdown` state. If it's already trying then it doesn't matter, the connectivity state will reflect that anyway.